File: python/src/linodemcp/tools/linode_profile_builder.py
"""Phase 8.2 read-only builder tools.

Two MCP tools enumerate the server's registerable tool surface for the
profile builder workflow:

- ``linode_profile_list_tools``: tool catalog with name, capability,
  and the list of categories each tool belongs to. Optional ``category``
  and ``capability`` filters.
- ``linode_profile_list_categories``: deduplicated category list with
  per-category tool counts, sorted by name.

Both tools carry ``Capability.Meta`` so the profile filter always
admits them, even under the read-only default profile. They never
touch the Linode API. The generator owns their registration; what is left
here is the body each one's answer hook calls.

The handlers read the live tool catalog off the builder state the server
publishes for the call (:mod:`linodemcp.tools.builderstate`), so a reload
reaches an already-registered tool and a handler reached without a server
refuses rather than answering from an empty catalog.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from mcp.types import TextContent

from linodemcp.genpb.linode.mcp.v1 import profile_builder_pb2
from linodemcp.profiles.builtin import categories as resolve_categories
from linodemcp.tools.builderstate import (
    BUILDER_UNCONFIGURED,
    builder_state_from_context,
)
from linodemcp.tools.helpers import error_response
from linodemcp.tools.proto_response import serialize_api_response

if TYPE_CHECKING:
    from linodemcp.profiles import Capability

# Argument-key constants. These are the JSON property names the model
# passes through MCP; hoisted so the handler and schema agree without
# stringly-typed drift.
_ARG_CATEGORY = "category"
_ARG_CAPABILITY = "capability"


def _capability_matches(capability: Capability, filter_value: str) -> bool:
    """Case-insensitive match against the long ("CapRead") or short ("read") form.

    Mirrors the Go ``capabilityMatches`` helper so cross-language tools
    accept the same filter strings.
    """
    long_form = capability.name  # Read, Write, Destroy, Admin, Meta
    cap_form = f"Cap{long_form}"

    needle = filter_value.lower()
    return needle in {long_form.lower(), cap_form.lower()}


def profile_list_tools_result(arguments: dict[str, Any]) -> list[TextContent]:
    """Return the filtered tool catalog as JSON text.

    The response shape per entry is ``{name, capability, categories}``
    where ``capability`` is the Capability stringified (``CapRead`` etc.)
    and ``categories`` is the list returned by
    :func:`linodemcp.profiles.builtin.categories`. Both filters narrow
    the result; missing/empty filters do nothing.
    """
    state = builder_state_from_context()
    if state is None:
        return error_response(BUILDER_UNCONFIGURED)

    category_filter = arguments.get(_ARG_CATEGORY, "")
    capability_filter = arguments.get(_ARG_CAPABILITY, "")

    entries = state.catalog()
    out: list[dict[str, Any]] = []

    for entry in entries:
        cats = resolve_categories(entry.name)
        if category_filter and category_filter not in cats:
            continue

        if capability_filter and not _capability_matches(
            entry.capability, capability_filter
        ):
            continue

        out.append(
            {
                "name": entry.name,
                "capability": f"Cap{entry.capability.name}",
                "categories": cats,
            }
        )

    # Name-sorted. The catalog itself is in registration order, which differs
    # per language, so sorting here is what makes one tool answer one order
    # whichever binary served it.
    out.sort(key=lambda entry: entry["name"])

    result = serialize_api_response(
        {"count": len(out), "tools": out},
        profile_builder_pb2.ProfileToolListResponse(),
    )
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


def profile_list_categories_result(arguments: dict[str, Any]) -> list[TextContent]:
    """Return ``[{name, tool_count}]`` for every category in the catalog.

    Counts include every category a tool carries (a tool that appears
    in two categories contributes 1 to each). Sorted by name so the
    output is reproducible and the cross-language parity test can
    compare directly.
    """
    # The tool takes no inputs per spec; del marks the parameter used.
    del arguments

    state = builder_state_from_context()
    if state is None:
        return error_response(BUILDER_UNCONFIGURED)

    entries = state.catalog()
    counts: dict[str, int] = {}

    for entry in entries:
        for cat in resolve_categories(entry.name):
            counts[cat] = counts.get(cat, 0) + 1

    out = [{"name": name, "tool_count": counts[name]} for name in sorted(counts)]

    result = serialize_api_response(
        {"count": len(out), "categories": out},
        profile_builder_pb2.ProfileCategoryListResponse(),
    )
    return [TextContent(type="text", text=json.dumps(result, indent=2))]

File: go/internal/tools/linode_profile_builder.go
package tools

import (
	"context"
	"slices"
	"sort"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
)

// ProfileListToolsAnswer answers linode_profile_list_tools. It enumerates
// every tool the server could register, independent of the active profile's
// filter, so the model can plan profile composition against the full menu.
// Optional filters trim the output:
//
//   - "category": include only tools that carry the named category
//   - "capability": include only tools tagged with the named capability
//     (one of CapRead, CapWrite, CapDestroy, CapAdmin, CapMeta)
//
// Both filters use exact match. "capability" checks are case-insensitive
// against the Capability String() form (the "Cap" prefix is optional, so
// "read" or "CapRead" both work).
//
// The answer is name-sorted. The catalog itself is in registration order,
// which differs per language, so sorting here is what makes one tool answer
// one order whichever binary served it.
func ProfileListToolsAnswer(
	ctx context.Context, request *mcp.CallToolRequest, _ *config.Config,
) (*mcp.CallToolResult, error) {
	state, refusal := builderStateOrRefusal(ctx)
	if refusal != nil {
		return refusal, nil
	}

	categoryFilter := request.GetString("category", "")
	capabilityFilter := request.GetString("capability", "")

	entries := state.Catalog()
	out := make([]*linodev1.ProfileToolCatalogItem, 0, len(entries))

	for idx := range entries {
		cats := profiles.Categories(entries[idx].Name)
		if categoryFilter != "" && !slices.Contains(cats, categoryFilter) {
			continue
		}

		if capabilityFilter != "" && !capabilityMatches(entries[idx].Capability, capabilityFilter) {
			continue
		}

		out = append(out, &linodev1.ProfileToolCatalogItem{
			Name:       entries[idx].Name,
			Capability: entries[idx].Capability.String(),
			Categories: cats,
		})
	}

	sort.Slice(out, func(i, j int) bool { return out[i].GetName() < out[j].GetName() })

	return MarshalProtoToolResponse(&linodev1.ProfileToolListResponse{
		Count: IDToInt32(len(out)),
		Tools: out,
	})
}

// ProfileListCategoriesAnswer answers linode_profile_list_categories. It
// reduces the catalog to the deduplicated category list with tool counts,
// sorted by name, so the model can discover what categories exist before
// drilling in with list_tools(category=...).
func ProfileListCategoriesAnswer(
	ctx context.Context, _ *mcp.CallToolRequest, _ *config.Config,
) (*mcp.CallToolResult, error) {
	state, refusal := builderStateOrRefusal(ctx)
	if refusal != nil {
		return refusal, nil
	}

	entries := state.Catalog()
	counts := make(map[string]int, 16)

	for idx := range entries {
		for _, cat := range profiles.Categories(entries[idx].Name) {
			counts[cat]++
		}
	}

	names := make([]string, 0, len(counts))
	for name := range counts {
		names = append(names, name)
	}

	sort.Strings(names)

	out := make([]*linodev1.ProfileCategoryItem, len(names))
	for i, name := range names {
		out[i] = &linodev1.ProfileCategoryItem{Name: name, ToolCount: IDToInt32(counts[name])}
	}

	return MarshalProtoToolResponse(&linodev1.ProfileCategoryListResponse{
		Count:      IDToInt32(len(out)),
		Categories: out,
	})
}

// capabilityMatches reports whether the given capability tag matches
// the filter string. Accepts both short ("read") and long ("CapRead")
// forms, case-insensitive, so the model can use either without an
// exact-string failure.
func capabilityMatches(capability profiles.Capability, filter string) bool {
	long := capability.String()
	if strings.EqualFold(long, filter) {
		return true
	}

	// "Cap" prefix is consistent across all Capability stringers, so
	// trimming it gives the short form. CapRead -> Read.
	const capPrefix = "Cap"
	if short, ok := strings.CutPrefix(long, capPrefix); ok && strings.EqualFold(short, filter) {
		return true
	}

	return false
}
